### PR TITLE
[SPARK-21954][SQL] JacksonUtils should verify MapType's value type instead of key type

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

`JacksonUtils.verifySchema` verifies if a data type can be converted to JSON. For `MapType`, it now verifies the key type. However, in `JacksonGenerator`, when converting a map to JSON, we only care about its values and create a writer for the values. The keys in a map are treated as strings by calling `toString` on the keys.

Thus, we should change `JacksonUtils.verifySchema` to verify the value type of `MapType`.

## How was this patch tested?

Added tests.

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/json/JacksonUtilsSuite.scala [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import java.util.TimeZone
+
+import org.apache.spark.SparkFunSuite
+import org.apache.spark.sql.RandomDataGenerator
+import org.apache.spark.sql.catalyst.{CatalystTypeConverters, InternalRow}
+import org.apache.spark.sql.types._
+
+class JacksonUtilsSuite extends SparkFunSuite {
+
+  test("verifySchema") {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/json/JacksonUtilsSuite.scala]: @viirya, would you mind if I ask to leave a simple e2e test alone for this PR? This one is quite straightforward to me and I think this won't require such many tests alone for this issue. I want to backport this but only leave strictly related changes here.

I think we could do something like `SELECT to_json(struct(map(interval 1 second, 'a')))` or `SELECT to_json(struct(map('a', interval 1 second)))`.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/json/JacksonUtilsSuite.scala [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      val gen = new JacksonGenerator(
+        schema, new CharArrayWriter(), new JSONOptions(Map.empty, TimeZone.getDefault.getID))
+      gen.write(row)
+    }
+
+    // The valid schema
+    val atomicTypes = DataTypeTestUtils.atomicTypes
+    val atomicArrayTypes = atomicTypes.map(ArrayType(_, containsNull = false))
+    val atomicMapTypes = for (keyType <- atomicTypes;
+      valueType <- atomicTypes) yield MapType(keyType, valueType, false)
```
Comment (HyukjinKwon) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/json/JacksonUtilsSuite.scala]: I'd do this as below:

```scala
val atomicMapTypes = for {
  keyType <- atomicTypes
  valueType <- atomicTypes
} yield MapType(keyType, valueType, false)
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/json/JacksonUtilsSuite.scala (hunk 1 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/json/JacksonUtilsSuite.scala]: OK, I am fine with those test coverage improvement too together. Up to you.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/json/JacksonUtilsSuite.scala [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+      val schema = StructType(StructField("a", dataType, nullable = false) :: Nil)
+      JacksonUtils.verifySchema(schema)
+      verfifyJSONGenerate(schema)
+    }
+
+    val invalidTypes = Seq(CalendarIntervalType)
+
+    // For MapType, its keys are treated as a string basically when generating JSON, so we only
+    // care if the values are valid for JSON.
+    val alsoValidMapTypes = for (keyType <- atomicTypes ++ invalidTypes;
```
Comment (HyukjinKwon) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/json/JacksonUtilsSuite.scala]: Would it be possible to move these supported cases into 45L so we loop twice, one for the vaild and one for invalid one?

**Comment by viirya** [comment]
@HyukjinKwon Thanks for review. I simplified the test cases. Please take a look when you are available.

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 4]
```diff
@@ -257,6 +257,18 @@ class JsonFunctionsSuite extends QueryTest with SharedSQLContext {
       "A type of keys and values in map() must be string, but got"))
   }
 
+  test("SPARK-21954: JacksonUtils should verify MapType's value type instead of key type") {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala]: @viirya, how about puting this test around `to_json unsupported type` here and maybe use Scala function API for consistency?

Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala]: Modified as suggestion.

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 5]
```diff
@@ -196,14 +196,28 @@ class JsonFunctionsSuite extends QueryTest with SharedSQLContext {
...
+
+    // interval type is invalid for converting to JSON. However, the keys of a map are treated
+    // as strings, so its type doesn't matter.
+    val df2 = baseDf
+      .select(struct(map($"a._1".cast(CalendarIntervalType), lit("a")).as("col1")).as("c"))
+    val df3 = baseDf
+      .select(struct(map(lit("a"), $"a._1".cast(CalendarIntervalType)).as("col1")).as("c"))
+    checkAnswer(
+      df2.select(to_json($"c")),
+      Row("""{"col1":{"interval -3 months 7 hours":"a"}}""") :: Nil)
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala]:
```scala
    val df2 = baseDf
      .select(struct(map($"a._1".cast(CalendarIntervalType), lit("a")).as("col1")).as("c"))
    ...
    checkAnswer(
      df2.select(to_json($"c")),
      Row("""{"col1":{"interval -3 months 7 hours":"a"}}""") :: Nil)
```

This case looks a supported case though. We could maybe make a separate test for this one.

**Review by HyukjinKwon — APPROVED** [review]
LGTM except for a minor comment.

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala (hunk 5 above)
Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala]: Ok. Moved into a test.

**Comment by HyukjinKwon** [comment]
Merged to master and branch-2.2.

**Comment by viirya** [comment]
Thanks @HyukjinKwon

**Pull request closed without merging**